### DISCUSSION TITLE
TypeScript compilation fails with "TS1479: The current file is a CommonJS module"

### DISCUSSION THREAD

**Discussion by BurningDog** [tailwindlabs/headlessui#2778] [Help] [answered]
Hi! First time using `@headlessui` but compilation is failing.  I'm trying out the [Vue Combobox](https://headlessui.com/vue/combobox). After `npm install @headlessui/vue` I've set up the `ComboBox` inside a Vue component. When I run `npm run watch` I get this error:

```
TS1479: The current file is a CommonJS module whose imports will produce 'require' calls; however, the referenced file is an ECMAScript module and cannot be imported with 'require'. Consider writing a dynamic 'import("@headlessui/vue")' call instead.
  To convert this file to an ECMAScript module, change its file extension to '.mts', or add the field `"type": "module"` to 'package.json'.
```

### The fix

To fix, I removed `"type": "module"` from `node_modules/@headlessui/vue/package.json` - I can now compile just fine and the combox box shows up in my browser, as expected.

I got that idea from https://github.com/tailwindlabs/headlessui/issues/1699#issuecomment-1190150520 which explores a related TypeScript compilation error - but that was fixed in https://github.com/tailwindlabs/headlessui/pull/1721 in July 2022.

Does anyone know why I am getting this error, and if there's another way to resolve it?

### My setup

I'm adding this to a [Symfony 6.2](https://symfony.com/) codebase, which uses [webpack-encore](https://github.com/symfony/webpack-encore), which is a light wrapper around `webpack`.

When running `npm run watch` (which is `encore dev --watch`) I get the above compilation error.

* vue version 3.0
* node version 18
* TypeScript version 4.9.5

In my `tsconfig.json` the relevant `compilerOptions` are:

```json
    "module": "ESNext",
    "moduleResolution": "NodeNext",
    "resolveJsonModule": true,
```

**Comment by djflorio** [comment]
I'm experiencing the same issue with the @headlessui/react library.

A potential workaround is to follow the instructions in the warning to dynamically import the components.

Instead of importing at the top of the file like this:
```
import { Listbox } from '@headlessui/react';

export const MyComponent = () => {
  return ...
}
```
Try async/awaiting the import within the component itself:
```
export const MyComponent = async () => {
  const { Listbox } = await import('@headlessui/react');
  return ...
}
```

This gets rid of the warning for me. I haven't been able to test it at runtime yet (in the middle of a big refactor), and I have no idea if this is an ideal solution, but it beats having to edit the headlessui package itself.

**Reply by djflorio** [reply]
Update: I finished my refactor and unfortunately this doesn't work for client components, which don't yet support async/await 😞

**Comment by greg-hoarau** [comment]
Hello,

I'm not sure about this, but you can use "moduleResolution" : "Bundler" instead of "NodeNext" with Typescript 5.

Good luck

**Reply by djflorio** [reply]
This worked for me! Note that you also need to make sure `module` is set to `es2015` or later. I'm using `module: ES2022` and `moduleResolution: Bundler` and that fixed the import issue.

**Reply by Sylvenas** [reply]
Thanks for your suggestion, that's great 🎉🎉

**Reply by MerryOscar** [reply]
This worked for me 🎉

**Reply by tinomazorodze** [reply]
Me too by changing ` "module": "NodeNext",` and `"moduleResolution": "NodeNext",` to `"module": "esnext",` and `"moduleResolution": "bundler"` the commonjs/ecmascript issue was fixed

**Reply by Pines-Cheng** [reply]
error TS6046: Argument for '--moduleResolution' option must be: 'node', 'classic', 'node16', 'nodenext'.

**Comment by BurningDog** [comment, marked as answer]
One of the maintainers realised that this was actually not working properly, and has a fix which will be in the next release. See https://github.com/tailwindlabs/headlessui/issues/2838#issuecomment-1852416904
